raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -194,14 +196,45 @@ private void completeCurrentBatch() {
         MemoryRecords data = currentBatch.build();
         completed.add(new CompletedBatch<>(
             currentBatch.baseOffset(),
-            currentBatch.records(),
+            Optional.of(currentBatch.records()),
             data,
             memoryPool,
             currentBatch.initialBuffer()
         ));
         currentBatch = null;
     }
 
+    public void appendLeaderChangeMessage(LeaderChangeMessage leaderChangeMessage, long currentTimeMs) {
```

Comment:
This method is not safe. I think you need to hold a lock before falling `maybeCompleteDrain` and updating `nextOffset`.